db/db_flush_test.cc

```diff
@@ -78,6 +80,62 @@ TEST_F(DBFlushTest, FlushWhileWritingManifest) {
 #endif  // ROCKSDB_LITE
 }
 
+// We had issue when a flush with wait triggered due to ExportColumnFamily can
+// finish before the result is committed to manifest. This is happening because
+// of a simultaneous flush on the column family. Run in loop to reproduce the
+// failure.
+TEST_F(DBFlushTest, FlushBeforeWritingManifestWithCheckpoint) {
+  Options options;
+  options.disable_auto_compactions = true;
+  options.max_background_flushes = 3;
+  options.max_write_buffer_number = 3;
+  env_->SetBackgroundThreads(3, Env::HIGH);
+  options.env = env_;
+  Reopen(options);
+
+  SyncPoint::GetInstance()->LoadDependency(
+      {{"DBImpl::FlushMemTable:AfterScheduleNonExportFlush",
+        "FlushJob::WriteLevel0Table:flush_started2"}});
+
+  SyncPoint::GetInstance()->SetCallBack(
+      "FlushJob::WriteLevel0Table:flush_started1", [&](void* /*arg*/) {
+        // Add second write and flush without wait.
+        FlushOptions no_wait;
+        no_wait.wait = false;
+        no_wait.allow_write_stall = true;
+        ASSERT_OK(Put("foo", "v"));
+        ASSERT_OK(dbfull()->Flush(no_wait));
+      });
+
+  SyncPoint::GetInstance()->SetCallBack(
+      "DBImpl::FlushMemTable:AfterScheduleNonExportFlush",
+      [&](void* /*arg*/) { SyncPoint::GetInstance()->DisableProcessing(); });
+
+  SyncPoint::GetInstance()->EnableProcessing();
+
+  rocksdb::ExportImportFilesMetaData* cf_sst_files_metadata = nullptr;
+  rocksdb::Checkpoint* checkpoint;
+  auto status = rocksdb::Checkpoint::Create(dbfull(), &checkpoint);
+  ASSERT_EQ(status.ok(), true);
+
+  // Add first write and trigger an ExportColumnFamily which will trigger a
```

Comment:
Export column family helps validate that we did commit some sst files. With just DB::Flush() I had no way of verifying.

Reply:
You need to verify your suspicion directly. Your suspicion is that, one thread issuing a manual flush returns prematurely, and this is the thing that this unit test must show. Ideally, the unit test should deterministically fail without your fix, and deterministically succeed with your fix. In this case, I checkout your code and neither is true. I also tried gtest-parallel, and it gave me a lot of other IO errors which may be related to export column family or the test not cleaning up the checkpoint dir path. Furthermore, exporting column family is a feature mush less rigorously tested and used than flush, thus I wouldn't rely on it to verify a more mature feature.

Adding a sleep via a syncpoint callback is not a fix, yet it does not repro either.

Reply:
I wrote a small script to run this test in a loop of 100. It fails significant number of times for me. This UT is mainly to show the issue, I have not fixed it in this pull request. My fix was to not populate flush_memtable_ids in https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_compaction_flush.cc#L2116 so that I only exit when NumNotFlushed is 0. I will try to write a separate test to demonstrate the num not flushed condition.

Reply:
This is not how we use unit test. A unit test must be able to reproduce deterministically with a single fresh run.
I also suspect what the root cause is in export column family. Since I am not familiar with that part, I would suggest that we use other approach to reproduce and show the bug directly, if it exists.

Reply:
Basically, the unit test need to directly demonstrate this
> Flush_B flushes its memtable before Flush_A and goes ahead with commit. This leads to Flush_A exiting the WaitForFlushMemTables before the commit to manifest happens.